cfy_manager/utils/certificates.py

```diff
@@ -259,67 +210,72 @@ def _generate_ssl_certificate(
     if isinstance(group, str):
         group = grp.getgrnam(group).gr_gid
 
-    # Remove duplicates from ips and ensure CN is in SANs
-    subject_altnames = _format_ips(ips, cn)
+    names = set()
+    for addr in itertools.chain(ips, [cn]):
+        if parsed_ip := network.parse_ip(addr):
+            names.add(x509.IPAddress(parsed_ip))
+            # don't break here: still add the IP as a DNS name:
+            # some browsers require that
+        names.add(x509.DNSName(addr))
 
     logger.debug(
         'Generating SSL certificate %s and key %s with subjectAltNames: %s',
-        cert_path, key_path, subject_altnames,
+        cert_path, key_path, names,
     )
 
-    csr_path = '{0}.csr'.format(cert_path)
+    key = rsa.generate_private_key(
+        public_exponent=65537,
+        key_size=4096,
+    )
+    subject = issuer = x509.Name([
+        x509.NameAttribute(x509.oid.NameOID.COMMON_NAME, cn),
+    ])
 
-    with _csr_config(cn, subject_altnames) as conf_path:
-        run([
-            'openssl', 'req',
-            '-newkey', 'rsa:2048',
-            '-nodes',
-            '-batch',
-            '-sha256',
-            '-config', conf_path,
-            '-out', csr_path,
-            '-keyout', key_path,
-        ])
-        if os.geteuid() == 0:
-            # Don't try to change cert/key ownership if we're not root
-            # (this indicates we're running non-sudo-commands such as
-            # generate-test-cert)
-            os.chown(key_path, owner, group)
-            os.chmod(key_path, key_perms)
-
-        x509_command = [
-            'openssl', 'x509',
-            '-days', '3650',
-            '-sha256',
-            '-req', '-in', csr_path,
-            '-out', cert_path,
-            '-extensions', 'v3_ext',
-            '-extfile', conf_path
-        ]
+    if sign_cert_path and sign_key_path:
+        with open(sign_key_path, 'rb') as sign_key_file:
+            sign_key = serialization.load_pem_private_key(
+                sign_key_file.read(),
+                password=sign_key_password.encode(),
+            )
+        with open(sign_cert_path, 'rb') as sign_cert_file:
+            sign_cert = x509.load_pem_x509_certificate(sign_cert_file.read())
+        issuer = sign_cert.subject
+    else:
+        sign_key = key
+
+    cert = (
+        x509.CertificateBuilder()
+        .subject_name(subject)
+        .issuer_name(issuer)
+        .public_key(key.public_key())
+        .serial_number(x509.random_serial_number())
+        # allow for up to 1 day of time drift, in case other nodes have
+        # unsynced clocks
+        .not_valid_before(datetime.utcnow() - timedelta(days=1))
+        # valid for 10 years
+        .not_valid_after(datetime.utcnow() + timedelta(days=3650))
```

Comment:
What is it with these 10 years?  What is the reason for creating certs like this?

Reply:
the direct reason is that the previous impl. did that :P 
We can change it, as long as the replace procedure is very smooth (which it's currently not, I think), but I'd prefer to avoid introducing behaviour changes in this PR